wgpu-core/src/command/mod.rs

```diff
@@ -78,21 +78,161 @@ pub(crate) enum CommandEncoderStatus {
     ///
     /// [`Global::queue_submit`] drops command buffers unless they are
     /// in this state.
-    Finished,
+    Finished(CommandBufferMutable),
 
     /// An error occurred while recording a compute or render pass.
     ///
     /// When a `CommandEncoder` is left in this state, we have also
     /// returned an error result from the function that encountered
     /// the problem. Future attempts to use the encoder (for example,
-    /// calls to [`CommandBufferMutable::check_recording`]) will also return
-    /// errors.
-    ///
-    /// Calling [`Global::command_encoder_finish`] in this state
-    /// discards the command buffer under construction.
+    /// calls to [`Self::record`]) will also return errors.
     Error,
 }
 
+/// Changes the state of the given encoder.
+///
+/// Takes `&mut CommandEncoderStatus`, the previous state kind and the next state kind.
+///
+/// Returns `&mut CommandBufferMutable`.
+///
+/// Will panic if the previous state is not accurate.
+macro_rules! change_state_with_inner {
+    ($self:ident, $before:path, $after:path) => {{
+        let v = ::core::mem::replace($self, CommandEncoderStatus::Error);
+        let $before(inner) = v else { unreachable!() };
+        let _ = ::core::mem::replace($self, $after(inner));
+        let $after(inner) = $self else { unreachable!() };
```

Comment:
Really? It seems necessary to me, since you've moved the `CommandBufferMutable` back into `$self`, and you need a mutable reference to it in situ. How would you do this better in modern Rust?

Reply:
There isn't a second destructuring of `$after(inner)` though, the first instance in `mem::replace($self, $after(inner))` is creating the `enum`.

Reply:
I ended up removing the macro.